// mailist/dist/structures/Client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_superfetch_1 = __importDefault(require("node-superfetch"));
const cheerio_1 = __importDefault(require("cheerio"));
const config = __importStar(require("../util/Constant"));
class Client {
    constructor(opt = {
        uri: "https://graphql.anilist.co/"
    }) {
        this.config = config;
        this.baseURL = opt.uri;
    }
    getAnime(query, hqlRequest) {
        return new Promise((resolve, reject) => {
            void node_superfetch_1.default.post(this.baseURL)
                .send(Object.assign({
                variables: {
                    search: query,
                    type: "ANIME"
                },
                query: hqlRequest
            }))
                .then(data => {
                return resolve(data.body);
            })
                .catch(reject);
        });
    }
    getManga(query, hqlRequest) {
        return new Promise((resolve, reject) => {
            void node_superfetch_1.default.post(this.baseURL)
                .send(Object.assign({
                variables: {
                    search: query,
                    type: "MANGA"
                },
                query: hqlRequest
            }))
                .then(data => {
                return resolve(data.body);
            })
                .catch(reject);
        });
    }
    getChar(query, hqlRequest) {
        return new Promise((resolve, reject) => {
            void node_superfetch_1.default.post(this.baseURL)
                .send(Object.assign({
                variables: {
                    search: query
                },
                query: hqlRequest
            }))
                .then(data => {
                return resolve(data.body);
            })
                .catch(reject);
        });
    }
    fetchMal(id) {
        return new Promise((resolve, reject) => {
            node_superfetch_1.default.get(`https://myanimelist.net/anime/${parseInt(id)}`)
                .then(text => {
                resolve(cheerio_1.default.load(text.text)("span[itemprop=\"ratingValue\"]").first().text());
            })
                .catch(reject);
        });
    }
}
exports.default = Client;
