// mailist/dist/getAnime.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_superfetch_1 = __importDefault(require("node-superfetch"));
const Client_1 = __importDefault(require("./structures/Client"));
class Anime {
    constructor() {
        this.client = new Client_1.default();
    }
    anime(query) {
        return new Promise((resolve, reject) => {
            this.client.getAnime(query, this.client.config.animeGraphql)
                .then(result => {
                if (!result.data.anime.results.length) {
                    throw new TypeError("Oh No, Anime Not Found");
                }
                resolve(result);
            })
                .catch(reject);
        });
    }
    Mal(id) {
        return this.client.fetchMal(id)
            .then(mal => mal);
    }
    voiceActors(query) {
        return __awaiter(this, void 0, void 0, function* () {
            const animeID = yield this.anime(query);
            if (!animeID.data.anime.results.length)
                throw new TypeError("Sorry, can't find that anime");
            return new Promise((resolve, reject) => {
                void node_superfetch_1.default.post(this.client.baseURL)
                    .send(Object.assign({
                    variables: {
                        id: animeID.data.anime.results[0].id
                    },
                    query: this.client.config.voiceActors
                }))
                    .then(data => {
                    return resolve(data.body);
                })
                    .catch(reject);
            });
        });
    }
}
exports.default = Anime;
